core/network/src/protocol/message.rs

```diff
@@ -284,16 +286,21 @@ pub mod generic {
 			if let Some(state) = &self.state {
 				state.encode_to(dest);
 			}
+			if let Some(data) = &self.data {
+				data.encode_to(dest)
+			}
 		}
 	}
 
 	impl<H: Decode> Decode for BlockAnnounce<H> {
 		fn decode<I: Input>(input: &mut I) -> Result<Self, codec::Error> {
 			let header = H::decode(input)?;
 			let state = BlockState::decode(input).ok();
+			let data = Vec::decode(input).ok();
```

Comment:
I'm not happy with this. Conceptually it should be `Option<(BlockState, Vec<u8>)>`. As both should be present with version 4. But yeah, for the sake of merging this and don't over-complicating it more, we probably can go with this.